### ISSUE TITLE
Allow usage of OpenShift versions lower than Minishfit's baseline

### ISSUE THREAD

**Issue by donovanmuller** [minishift/minishift#667]
Just tried `minishift-1.0.0-rc.1` with an _existing_ OpenShift 1.4.1 instance (created with `minishift-1.0.0-beta.5`) and when trying to start, I get this message:

```console
$ ./minishift start --cpus 4 --memory 8192 --disk-size 30g --openshift-version v1.4.1
Minishift does not support Openshift version v1.4.1 .You need to use a version >=v1.5.0-rc.0
```

Is there any reason why OpenShift 1.4.x wouldn't/shouldn't be supported by `minishift` > 1.0.0?

I don't want to `delete` my existing instance but I'm assuming I would get the same message if I created a new instance but targeted at 1.4.1...

**Comment by praveenkumar** [comment]
@donovanmuller We took this step because there are lot of regression/features which we implemented as part of rc1 release only supported by `v1.5.0-rc0` and not back-ported by older version. If you want to use older openshift version then please use older bit of minishift which works.

**Comment by budhrg** [comment]
> If you want to use older openshift version then please use older bit of minishift which works.

@donovanmuller, specifically  `v1.0.0-beta.5` or below supports `1.4.1`.

**Comment by hferentschik** [comment]
Technically one could argue that the version check comes a bit too early. Given that the VM exist, provisioning would not happen anyways. The check for version should probably be after the check for existing API.

**Comment by LalatenduMohanty** [comment]
@donovanmuller Is there any specific reason you want to stick to version 1.4.1? Because we assumed that users would be more interested to go for newer version of OpenShift.

**Comment by kadel** [comment]
> @donovanmuller Is there any specific reason you want to stick to version 1.4.1? Because we assumed that users would be more interested to go for newer version of OpenShift.

I've also hit this issue. I want to use same version of OpenShift in minishift as is in my production environment.
People will run 1.4 in production for some time.

**Comment by donovanmuller** [comment]
@LalatenduMohanty @kadel I also want to mirror our production OpenShift 3.4 installation, which is the main reason for asking about 1.4/3.4 compatibility.

We will move to OpenShift 3.5 when it's released but it would be be nice to have version parity between local development environments and what's running in production.

@praveenkumar Just out of interest, what are the changes that cannot be applied to < 1.4?

**Comment by praveenkumar** [comment]
> Just out of interest, what are the changes that cannot be applied to < 1.4?

Proxy support which is included in 1.5 (oc cluster up) and another one insecure registry (again added to 1.5).

**Comment by donovanmuller** [comment]
👍

**Comment by jorgemoralespou** [comment]
@praveenkumar proxy support was backported to 1.4.1, not sure what fix you talk about for insecure-registry IP.

I would also consider this a requirement, and possibly not bootstrap things not supported by those versions. As @donovanmuller mentions and @kadel, people will want to use same cluster in minishift as they are using in their production.

**Comment by hferentschik** [comment]
@jorgemoralespou, OpenShift 1.5 has a fixed IP for the registry, which allows you to push directly to this registry. This relates to the exposing of the registry as route issue - #277. As you know there are already a lot of moving parts here. The idea off not allowing any OpenShift version lower than the one we use per default, is to reduce the amount of variables we have to deal with when dealing with user requests. 

Excluding 1.4 might have been a little bit too fast, in what you guys describe. We can discuss this once more.

**Comment by jorgemoralespou** [comment]
@hferentschik I still not sure what has changed in OpenShift or "oc cluster" related to the registry, I might have missed that.

@csrwng are you aware of any big change that would prevent running oc 1.4 in minishift apart from those 2 described? Although I'm sure proxies is there already in 1.4.1.

**Comment by csrwng** [comment]
@jorgemoralespou proxy support was backported to ocp 3.4.1, but not origin 1.4.1.

**Comment by jorgemoralespou** [comment]
Damn it.

**Comment by LalatenduMohanty** [comment]
@jorgemoralespou as you can expect once we reach GA we will be extra/more careful about maintaining the backward compatibility with OpenShift versions.

**Comment by LalatenduMohanty** [comment]
> Excluding 1.4 might have been a little bit too fast, in what you guys describe. We can discuss this once more.
+1

**Comment by hferentschik** [comment]
Here is what I suggest. The thing which gets into way is just the minimum version check. We know the basic provisioning works. 

So what's about we change the code the following way:

* Make the minimum version 1.4.1 
* Add another check which checks whether the requested version (as per _minishift start) is lower than the default version we are aligning with. In this case we print a warning message, saying that "Version x.y.z is lower that the currently supported default version u.v.w. Provisioning might fail and not all features might work as expected." (something along these lines)
* In the docs we add a note, that we are not supporting any version lower than the current default. Provisioning might work, but we are not supporting it in case something does not work.

Thoughts?

**Comment by jorgemoralespou** [comment]
@hferentschik I'm all up to support previous versions (even 1.3). On the
CDK front you'll get this request quite soon, as there is still many
customers on 1.3 (I understand not supporting 1.2).

Also, another option, which is very handy, and resembles what I do in
oc-cluster could be to use the 1.5.0 client in the VM, but pull down the
1.4 or 1.3 images. This works like a charm, and you can get proxy
configured on a 1.3 cluster.

**Comment by hferentschik** [comment]
> CDK front you'll get this request quite soon, as there is still many customers on 1.3 (I understand not supporting 1.2).

Right, we will discuss this next week and take some action. As said, I am happy to allow attempts to install earlier versions, with the caveat that one is really on his own there.

> Also, another option, which is very handy, and resembles what I do in oc-cluster could be to use the 1.5.0 client in the VM, but pull down the 1.4 or 1.3 images. This works like a charm, and you can get proxy configured on a 1.3 cluster.

That's an interesting thought and would be quite easy to do for us. It is a bit against the idea to always align _oc_ version with the cluster version you are running, but we could maybe say to align _oc_ version if OpenShift is >= default OpenShift version, but use the default version were earlier versions of OpenShift are to be provisioned. That's an interesting thought and is somewhere a middle ground. 

Just to be sure, all the code needed for cluster up is literally in the _oc_ binary. There is nothing backed in into the images, right?

**Comment by csrwng** [comment]
> Just to be sure, all the code needed for cluster up is literally in the oc binary. There is nothing backed in into the images, right?

yes, that is correct

**Comment by hferentschik** [comment]
Sweet

**Comment by hferentschik** [comment]
Merged via pull request #778
